### DISCUSSION TITLE
Introduce kata without a reward for solving

### DISCUSSION THREAD

**Discussion by hobovsky** [codewars/codewars.com#2568] [Ideas]
An idea inspired by kata like [Get The Substring](https://www.codewars.com/kata/61e128dc748f97004b46ea0d/discuss#61e2363f4f14bb0057dd8f3b): introduce a category of kata which, upon solving, would grant no reward.

Such kata could be used to introduce tasks which usually ask to reimplement a built-in functionality which seems to be interesting from some point of view (for example algorithmic), but do not make good Codewars kata. With no reward, the kata would not be prone to abuse by working around the code restrictions, and would hopefully attract only users who actually want to implement a requested solution rather than just call a single line.

Such (let's call them) "9 kyu" kata would be useful for:
- tasks like "implement KMP algorithm for matching a substring in a string"
- those stupid golf tasks like "`Math` fell out of your computer and now you need to reimplement `sin` and `log` without using any arithmetic operators, with code below 42 characters" ([example](https://www.codewars.com/kata/61e088c6e8c73e000f9fdda9))
- Big integers arithmetic kata for languages with built in big int support (i.e. more and more of them)
- ... and basically other tasks which attempt to enforce difficulty by restrictions on source code and usually fail miserably doing so.

What do you think?

**Comment by JohanWiltink** [comment]
I [have](https://www.codewars.com/kata/58638bd2210be9a9690001f7) a prime candidate for that.

I like the idea. A lot.

Can we disable upvoting solutions for that too? ( cq. have it giving no honour? )

**Comment by Blind4Basics** [comment]
I... Don't like the idea.

For builtins, maybe, but stuff like "implementing KMP", definitely no. Imo, going that way leads to "if you can find a solution described somewhere, this is 9 kyu". That means it's already appliable to... dunno... half of the kata database? (yeah, less than this, but you get the idea)

Reading about the BigInt part of the original message, I'm starting to feel like this is actually an idea to avoid retirement/unpublishing of problematic kata. That sounds quite different from just "suggesting a new 9kyu rank".

Let's imagine it's created and integrated to all other ranks.
- users can use it to rank kata ? => even more mess than before in the ranking, as long as the rank scale stays so blurry. Ppl will start to use it as actual rank, trolls will use it too (yeah, I have some specific users in mind...)
- or is that a mod feature?

**Reply by hobovsky** [reply]
> For builtins, maybe, but stuff like "implementing KMP", definitely no. Imo, going that way leads to "if you can find a solution described somewhere, this is 9 kyu". That means it's already appliable to... dunno... half of the kata database? (yeah, less than this, but you get the idea)

The KMP example is not there because its something what is easily searchable, but because its already available in string libraries as `indexOf`. So basically now you cannot have kata about substring search algos, or sorting algos, or arithmetic algos,  because they can be solved with built-in tools. But I think the algos themselves are very interesting and worth training.

**Reply by Blind4Basics** [reply]
oh, good point.
(Unless the restriction are succesfully enforced?)

**Reply by hobovsky** [reply]
> Unless the restriction are succesfully enforced?

They never are. Even if they are in one language, then a translation into another language will pop up, where its not possible to block.

Additionally, except maybe the "stupid golf tasks", such problems are usually not related to a specific language. Its not really good to have a bigint kata in js because built in bigint can be sufficiently blocked, but not have it in python only because its difficult to restrict there.

**Reply by Blind4Basics** [reply]
yeah, true.

**Comment by JohanWiltink** [comment]
It should _not_ be integrated with the other ranks.

But I think it would be useful for ~~kata~~ _tasks_ that focus on process rather than result.

**Comment by Blind4Basics** [comment]
It definitely feels like it's not a ranking for a kata, but something else.

At this point I'm like "why not", but I'd suggest using something else than "kyu/dan". Internally, that would work like a kyu/dan thing (with 0 values), but the users should imo see that as something completely different. No idea what to use (someone specialist in martial arts?), maybe soemthing about training. What that's thing karate guys do, when they repeat over and over the very same.... "succession of figures"? (I don't find my words, there)

**Reply by hobovsky** [reply]
I wanted to avoid mentioning the kyu thing.  Its not in the title, but then it slipped through in the text, probably causing confusion.

**Reply by Blind4Basics** [reply]
(I was coming from discord, and that was in the title there x) 'didn't even see it wasn't, here)

**Reply by Blind4Basics** [reply]
(I wanted to put a :+1:  on your message but there is a weird UI bug that forbids me to do it, now that I've already posted an answer... :/ )

**Comment by Kacarott** [comment]
I agree that it should be kept seperate from the rank system. Perhaps seperate "ranked" katas from "unranked" ones, and rank/honor/points can only be gained from ranked ones. (This kind of system is fairly common in video games, etc so will likely be intuitive to many users). Perhaps in "unranked", katas can have a colour (white/yellow/blue/purple) to be roughly indicative of difficulty, but no number attached.

**Reply by Kacarott** [reply]
In this regard, I imagine it as an internal kata setting the author can choose when creating the kata. (And mods would also be able to revert ranked katas to unranked if necessary). NOT something that beta reviewers can vote on. Perhaps if a beta is unsuitable for ranked, solvers can raise issues suggesting setting to unranked, but I think that should be up to authors/mods to actually set it.

**Comment by Blind4Basics** [comment]
note: if it's used to handle problematic kata, that suggests their "rank" will be changed, meaning shittons of server loads, to recompute the honor and rank for most of the users databse (since they are the most completed kata, generally).
So for old kata, it still could be safer to just retire them (or remove the problematic translations if possible?), and use that new classification only for new kata.

**Reply by Kacarott** [reply]
This could probably be decided on a per-kata basis. We could make a list of katas to deal with and kazk could let us know which ones look feasible to convert?

**Comment by Kacarott** [comment]
I've had another idea related to this recently. For certain events (for example, Advent of Code) many people enjoy sharing their solutions with each other in a variety of languages, and seeing how other people solved the same problem.

If unranked kata were implemented, then they could be set up with all available languages, and with no tests (so any code can be submitted) to act as somewhere for people to share their solutions with each other. This would be a far more attractive place to share solutions since on a Codewars kata there is syntax highlighting, and you can easily sort by language and/or users you follow, which is far better than many other places people often share code, such as reddit or discord.

I think this would open up a lot of opportunities for fun community "events" for people to get involved in.

*Finally, just to be completely clear, I am not suggesting that we would copy AoC (or other events) challenges, we would just be setting up an unranked kata as a place to share solutions. I imagine the description of such a kata would be little more than a link to the actual challenge page.*

**Comment by JohanWiltink** [comment]
So, a kumite?

**Reply by Kacarott** [reply]
Well, no, because a kumite requires a new fork for every new solution. You also cannot easily sort children of a kumite by solver. An actual kata provides a much better interface to see many solutions at once, and to sort/filter them. In general, Kumite is better for 1 on 1 comparison, a Kata is much better for larger groups of solutions.

**Reply by JohanWiltink** [reply]
A different UI, more like kata, over kumite then? Would that work?

( Just trying to see where this would fit in really. )

**Reply by Kacarott** [reply]
Potentially? Though I think just using the Kata format is the easiest. The only difference would be that no points would be awarded, and that for "Special events" like AoC, no tests would run either. But the solution page of a kata seems ideal

**Reply by JohanWiltink** [reply]
Having the _option_ to run tests, but not having to pass them, seems like an advantage.

( Can always not have tests, except for an always passing one )

**Comment by awesomeAD1** [comment]
Sorry, I'm late. Just stumbled on this. I like the idea, but this worries me:
> Big integers arithmetic kata for languages with built in big int support (i.e. more and more of them)

Presumably, such a kata would not bother enforcing restrictions (because it's impossible), meaning I could simply let the BigInt-implementing language solve this trivially, then get access to "proper" solutions, then use said solutions to solve the actual (ranked) kata where the restrictions are in place (or in a language without voluminous ~~whale~~ whole numbers)

**Reply by hobovsky** [reply]
Oh that's a good point, I did not think of it when writing the cited post.
So basically the problem is that having an unranked bigint kata in Python, and an equivalent ranked kata in C, solutions visible after solving the Python version can be translated to C and spoil a potentially rewarding kata.

Definitely something to consider.
